File: benchmarks/fixtures/webapp_py/middleware/logging_mw.py
"""Request/response logging middleware."""

import time
from typing import Any, Dict

from ..utils.logging import get_logger
from ..utils.helpers import validate_request, generate_request_id, mask_sensitive

_logger = get_logger("middleware.logging")

SENSITIVE_FIELDS = ["password", "token", "secret", "api_key", "authorization"]


def logging_middleware(request: Dict[str, Any]) -> Dict[str, Any]:
    """Log incoming requests with timing and request ID."""
    validate_request(request)

    # Assign request ID
    request_id = request.get("request_id") or generate_request_id()
    request["request_id"] = request_id

    # Log sanitized request
    safe_request = mask_sensitive(request, SENSITIVE_FIELDS)
    method = request.get("method", "?")
    path = request.get("path", "?")
    _logger.info(f"[{request_id}] {method} {path}")

    # Record timing
    request["_start_time"] = time.time()
    return request


def log_response(request: Dict[str, Any], status: int, body_size: int = 0) -> None:
    """Log response details with timing."""
    request_id = request.get("request_id", "unknown")
    start = request.get("_start_time", time.time())
    duration = (time.time() - start) * 1000  # milliseconds

    method = request.get("method", "?")
    path = request.get("path", "?")
    _logger.info(f"[{request_id}] {method} {path} -> {status} ({duration:.1f}ms, {body_size}B)")

File: benchmarks/fixtures/webapp_rs/middleware/logging_mw.rs
use crate::utils::helpers::{get_logger, generate_request_id};
use crate::Request;
use crate::Response;

/// Log an incoming request with method, path, and generated request ID.
pub fn log_request(request: &Request) -> String {
    let logger = get_logger("middleware.logging");
    let request_id = generate_request_id();
    logger.info(&format!(
        "[{}] {} {}",
        request_id,
        "GET",
        request.path,
    ));
    request_id
}

/// Log an outgoing response with status and request ID.
pub fn log_response(request_id: &str, response: &Response) {
    let logger = get_logger("middleware.logging");
    logger.info(&format!(
        "[{}] Response: status={}",
        request_id,
        response.status,
    ));
}

/// Log an error that occurred during request processing.
pub fn log_error(request_id: &str, error: &str) {
    let logger = get_logger("middleware.logging");
    logger.error(&format!("[{}] Error: {}", request_id, error));
}

/// Middleware that wraps a handler with request/response logging.
pub fn with_logging(
    request: &Request,
    handler: fn(&Request) -> Response,
) -> Response {
    let logger = get_logger("middleware.logging");
    let request_id = log_request(request);
    logger.info(&format!("[{}] Processing request", request_id));
    let response = handler(request);
    log_response(&request_id, &response);
    response
}
